Mark list for regions

PeterSolMS opened the pull request:
This PR adapts the mark list logic for regions.

The basic idea is that as plan_phase goes through the regions, it calls get_region_mark_list obtain the mark list for each region. 

How get_region_mark_list works is different for WKS and SVR flavors of the GC:

- for WKS, we have a single sorted mark list. Finding the mark list section corresponding to a region is done via binary search in the mark list.

- for SVR, each GC thread, after sorting the mark list, determines where the sections corresponding to all the regions start and end. This is stored in tables mark_list_piece_start and mark_list_piece_end which are now indexed by region number. Later, get_region_mark_list merges the pieces for a region from all heaps. This is very similar to what merge_mark_lists did before.

When reviewing, ignore the changes to the handle table code - this has bug fixes from Maoni that she will merge before me.

PeterSolMS committed:
delete backport.yml so I can push to master

PeterSolMS committed:
delete backport.yml so I can push to master

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Undo unintended changes

PeterSolMS committed:
Undo unintended changes

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge from upstream/master

PeterSolMS committed:
Merge from upstream/master

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge branch 'master' of https://github.com/PeterSolMS/runtime-1

PeterSolMS committed:
Undo unintended Mono changes that somehow crept in.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Check point for mark list implementation for regions - not yet in a working state, but hopefully close for SVR. For WKS, get_region_mark_list needs to be implemented.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge branch 'master' into mark_list_for_regions

PeterSolMS committed:
Implement get_region_mark_list for WKS.

The implementation just does a binary search for the start and end of the region in the mark list. This should require O(log(#mark list items)) and thus should be significantly cheaper than sorting the mark list items. So I think this should be fast enough. If not, there are faster implementations (but using more memory most likely).

PeterSolMS committed:
Small bug fixes:
- Fix #ifdef so mark list related code is only included if MARK_LIST is defined
- Fix calls to get_region_mark_list so that it writes the end to a local variable mark_list_index instead of mark_list_end.
- Fix from Maoni in GCHeap::IsPromoted

PeterSolMS committed:
Quick hack to set the range in sort_mark_list (tighter bound will need more work), bug fix from Maoni in  process_last_np_surv_region.

PeterSolMS committed:
Get better bounds for the range of the mark list items in sort_mark_list. Attempted bug fix for empty regions in relocate_survivors and compact_phase.

PeterSolMS committed:
Small fixes:
- #define STRESS_REGIONS again
- fix computations at the start of sort_mark_list for regions.
- added assert to detect sort range issues before attempting to sort
- go to COOP mode to assign handles in STRESS_REGIONS code
- disable code to create pins for STRESS_REGIONS except for clrgc.dll (we need to make this better)
- fix to set the allocation segment in process_last_np_surv_region for the case where we needed to create new region. Need to restructure the code so we don't have as much code duplication.

PeterSolMS committed:
Restructure code in gc_heap::process_last_np_surv_region to reduce code duplication.

PeterSolMS committed:
Minor changes to whitespace, parentheses etc., incorporated Maoni's changes to the handle table code so that coreclr.dll now works with STRESS_REGIONS.

PeterSolMS committed:
Disable USE_REGIONS for submitting the PR.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -9349,6 +9367,60 @@ size_t gc_heap::sort_mark_list()
+        delete[] mark_list_piece_start;
+        delete[] mark_list_piece_end;
+
+        mark_list_piece_start = new (nothrow) uint8_t** [region_count];
+        mark_list_piece_end   = new (nothrow) uint8_t** [region_count];
```
this is definitely not urgent (because we don't even change the lowest/highest address right now) but in the future when we do we'd want some kind of growth factor instead of allocating the exact region_count elements...basically like what we do with all the other data structures when we grow.

PeterSolMS replied:
Makes sense - done.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -9349,6 +9367,60 @@ size_t gc_heap::sort_mark_list()
+            if (mark_list_piece_end)
+                delete[] mark_list_piece_end;
+            mark_list_piece_size = 0;
+            return 0;
```
we'd want to fake a mark list overflow here as well before returning.

PeterSolMS replied:
Right, the failure case handling is incomplete. I think faking a mark list overflow doesn't quite do it, rather I'll need to check mark_list_piece_size for *all* the heaps in plan_phase when we compute use_mark_list.

Maoni0 replied:
what I meant was to fake a mark list overflow like you did with this code toward the beginning of sort_mark_list so you didn't need to have plan phase check for the mark_list_piece_size 

```cplusplus
    // if any other heap had a mark list overflow, we fake one too,
    // so we don't use an incomplete mark list by mistake
```
however that's also not correct 'cause some heaps are already using their mark_list_index so it'd be wrong to change it. but now I've thought about it a bit more I think we could make this better by allocating the memory for the new mark_list_piece*, if needed, before sort_mark_list. so for regions we'd have something like grow_mark_list_pieces before the `gc_join_null_dead_long_weak` join. and in the join we fake a global mark list overflow if any heap failed to alloc memory for their mark_list_piece*. this way we don't need to call sort_mark_list at all on any heap if we already don't have memory for the pieces (so we save the sort - of course you can achieve that part by simply moving the alloc code toward the top in sort_mark_list for that heap; but this way we save it for all heaps). what do you think? am I missing anything?

another thing we can do to optimize is to get the actual region count but that can be part of the server GC work.

PeterSolMS replied:
Well, perhaps it would be best to do the allocation at the very start of the mark phase and disable the mark list if we don't succeed, similarly to how we disable it for gen 2 GCs?

Maoni0 replied:
that's fine too...doesn't make a big difference to me.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -9625,7 +9708,56 @@ void gc_heap::merge_mark_lists (size_t total_mark_list_size)
+    // do a binary search over the sorted marked list to find start and end of the
+    // mark list for this region
+    *mark_list_end_ptr = binary_search (mark_list, mark_list_index, end);
+    return binary_search (mark_list, mark_list_index, start);
```
you could use mark_list_end_ptr you get in the 1st binary_search in the 2nd one 'cause we know we don't need to search anything between it and mark_list_index.

PeterSolMS replied:
Yes, this should shave off about one iteration on average - done.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -9472,13 +9543,15 @@ static int __cdecl cmp_mark_list_item (const void* vkey, const void* vdatum)
+#ifdef USE_REGIONS
+uint8_t** gc_heap::get_region_mark_list (uint8_t* start, uint8_t* end, uint8_t*** mark_list_end_ptr)
+{
+    size_t lowest_region_number = (size_t)g_gc_lowest_address >> gc_heap::min_segment_size_shr;
+    size_t region_number = ((size_t)start >> gc_heap::min_segment_size_shr) - lowest_region_number;
```
nit - would be nice to make a small helper get_region_index method for something like this - I'm sure we'll be using it in the future work.

PeterSolMS replied:
I implemented a helper get_basic_region_index_for_address - it computes the region_number as in the second line in the code you quoted, i.e. subtracts lowest_region_number.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -18399,7 +18550,7 @@ void gc_heap::init_background_gc ()
 
     //reset the plan allocation for each segment
-    for (heap_segment* seg = generation_allocation_segment (gen); seg != ephemeral_heap_segment;
+    for (heap_segment* seg = generation_allocation_segment (gen); seg != nullptr && seg != ephemeral_heap_segment;
```
you don't need this change anymore - my last checkin put this code under #ifndef USE_REGIONS.

PeterSolMS replied:
Ok - done.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -14605,6 +14755,8 @@ BOOL gc_heap::soh_try_fit (int gen_number,
 #ifdef USE_REGIONS
                 if (can_allocate)
                 {
+                    // disable for GC in coreclr, as we run into an assert there because
```
changes in this method can be eliminated if you just merge with the latest.

PeterSolMS replied:
Right, done.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -24695,6 +24846,14 @@ void gc_heap::process_last_np_surv_region (generation* consing_gen,
                     heap_number, heap_segment_gen_num (next_region), heap_segment_mem (next_region)));
                 assert (next_region != 0);
             }
+            else if (settings.promotion)
+            {
+                assert(next_plan_gen_num == 0);
+                next_region = get_new_region(0);
+
+                dprintf(REGIONS_LOG, ("h%d getting a new region for gen0 plan start seg to %Ix",
+                    heap_number, heap_segment_mem(next_region)));
+            }
```
changes in this method were also included with the lastest checkin

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -28200,7 +28376,14 @@ void gc_heap::relocate_survivors (int condemned_gen_number,
                     current_heap_segment = heap_segment_next_rw (current_heap_segment);
                     current_brick = brick_of (heap_segment_mem (current_heap_segment));
-                    end_brick = brick_of (heap_segment_allocated (current_heap_segment)-1);
+                    end_brick = brick_of (heap_segment_allocated (current_heap_segment) - 1);
+                    if (heap_segment_mem (current_heap_segment) == heap_segment_allocated (current_heap_segment))
+                    {
+                        // the case of an empty region / segment - make sure
+                        // we don't relocate anything, but go to the next region / segment
+                        assert (current_brick == end_brick);
+                        current_brick++;
+                    }
```
this can also be eliminated

PeterSolMS replied:
Why is that? Is there another change in your latest checkin that makes sure we never encounter an empty region here?

PeterSolMS replied:
As it turns out, the issue why I added this code is still there. However, my assert was wrong, so I removed it.

Maoni0 replied:
hmm, that's interesting. I had thought you only needed this because of the bug I fixed in the last checkin. it certainly should be able to handle empty segments or we would have hit this many years ago. so at the very least I don't think this is a problem for segments. I cannot repro this with regions either. the first brick should be -1 (constructed by the plan phase). how did you repro this? could you show the exact command line, the env vars you set and how you built (what features you included)?

PeterSolMS replied:
My repro is with server GC, using the debug build of the master branch of a couple days ago (of course with USE_REGIONS defined, but without any of mymark list changes). Command line is this (copied from your checkin comments for the regions work):

c:\Debuggers\windbg.exe C:\runtime-1\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\CoreRun.exe GcPerfSim.dll -tc 2 -tagb 16 -tlgb 0.05 -lohar 0 -sohsi 10 -lohsi 0 -pohsi 0 -sohpi 0 -lohpi 0 -pohpi 0 -sohfi 0 -lohfi 0 -pohfi 0 -allocType reference -testKind time

Environment variables are these:

I:\GCPerfSim>set complus
complus_gcConcurrent=0
complus_GCRegionsRange=800000000
COMPLUS_gcServer=1

Local variables in relocate_survivors are these:

0:013> dv
              end_brick = 0x127800
                   args = struct SVR::gc_heap::relocate_args
          start_address = 0x000001bd`6e000020 ""
            end_address = 0x000001bd`6e3ff7a8 ""
          condemned_gen = 0x000001bc`c61de6a8
   current_heap_segment = 0x000001c4`c87ea430
          current_brick = 0x127800
                      i = 0n0
                   this = 0x000001bc`c61de550
   condemned_gen_number = 0n2
first_condemned_address = 0x000001bc`e7c00020 "???"
           stop_gen_idx = 0n0

and brick_table[current_brick] is 0.

I think that empty segments work fine without regions, because heap_segment_mem is normally brick aligned, so subtracting one from heap_segment_allocated will result in a smaller brick if the two are the same. With regions, heap_segment_mem is not brick aligned, so subtracting one results in the same brick, which is what I observe.

Maoni0 replied:
but my point was the brick that corresponds to heap_segment_mem (not (heap_segment_mem - 1)) should be -1, not 0. so it's puzzling why this is 0 for you. I've been running this for a while and cannot repro (I've tried with and without STRESS_REGIONS).  also end_address looks weird - it should be the same as start_address for empty segs. I'm not sure if it's just not displaying the value correctly. I'll see if I can catch you online later today to debug this.

Maoni0 replied:
oh actually I did mean to mention this - you don't need this anymore with the change in get_free_region, right?

PeterSolMS replied:
Correct - I verified that the change in get_free_region fixes it.

PeterSolMS replied:
And yes, I should take out this code if it's no longer needed. Will double check...

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -29124,6 +29307,13 @@ void gc_heap::compact_phase (int condemned_gen_number,
                     {
                         args.src_gennum = ((current_heap_segment == ephemeral_heap_segment) ? -1 : 2);
                     }
+                    if (heap_segment_mem (current_heap_segment) == heap_segment_allocated (current_heap_segment))
+                    {
+                        // the case of an empty region / segment - make sure
+                        // we don't compact anything, but go to the next region / segment
+                        assert (current_brick == end_brick);
+                        current_brick++;
+                    }
```
same with this.

PeterSolMS replied:
I think my comments above apply to this location as well.

Maoni0 commented:
@PeterSolMS sorry you'd have to sync to the latest src to resolve some merge conflicts :)

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge with master.

PeterSolMS committed:
Addressed code review feedback:
- new helper method get_basic_region_index_for_address
- fix error handling for the case where mark_list_piece_start/mark_list_piece_end cannot be allocated
- small optimization in WKS::gc_heap::get_region_mark_list: limit range for the second binary search based on the result from the first binary search.

PeterSolMS committed:
Fix build issue, remove unintentional blank line.

PeterSolMS committed:
Removed erroneous assert, fixed whitespace changed by mistake.

PeterSolMS committed:
Move initialization of mark_list_piece_start/end to beginning of mark_phase and allocate it as one big array - that way it's easier to handle the case when allocation of the arrays fails. In this case, we just disable the mark list like we do for a blocking gen 2.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge branch 'master' into mark_list_for_regions

PeterSolMS committed:
Fix result of bad merge where we ended up with one #endif too many.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -22829,6 +22977,23 @@ void gc_heap::mark_phase (int condemned_gen_number, BOOL mark_only_p)
+            // disable use of mark list altogether
+            mark_list_piece_start = nullptr;
+            mark_list_piece_end = nullptr;
+            mark_list_end = &mark_list[0];
```
this would mean that you will set `mark_list_overflow` to true in `sort_mark_list` even though you don't need to grow, wouldn't it?

PeterSolMS replied:
Ah, right, great catch - thanks! I will test for the case of g_mark_list_piece being null at the start of sort_mark_list and treat it the same as a gen 2.

Maoni0 reviewed (approved):
other than [this comment](https://github.com/dotnet/runtime/pull/46979#pullrequestreview-576931750) it LGTM.

PeterSolMS committed:
Fix logic error with error case handling for allocation of g_mark_list_piece found in code review.

When we cannot allocate g_mark_list_piece, we disable the mark list. We need to guard against interpreting this as a mark list overflow, which would cause us to attempt to grow the mark list.

PeterSolMS committed:
Remove bug fixes in relocate_survivors and compact_phase that are no longer needed because get_free_region initializes the first brick.
